runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5278,6 +5287,40 @@ class INTERPRETER_CLASS
 			} else if (J9NtcPointer == argType) {
 				/* ffi_call expects the address of the pointer is the address of the stackslot. */
 				pointerValues[i] = (U_64)ffiArgs[i];
+#if JAVA_SPEC_VERSION >= 22
+				if (J9_ARE_ALL_BITS_SET(pointerValues[i], J9_FFI_DOWNCALL_HEAP_ARGUMENT_ID)) {
+					U_64 heapOffset = *(U_64 *)J9JAVAARRAY_EA(_currentThread,
+							J9_JNI_UNWRAP_REFERENCE(_sp + 10), /* The heap offset array */
+							curHeapArgIdx,
+							U_64);
+					j9object_t heapBase = (j9object_t)J9JAVAARRAYOFOBJECT_LOAD(_currentThread,
+							J9_JNI_UNWRAP_REFERENCE(_sp + 11), /* The heap base array */
+							curHeapArgIdx);
+					if (NULL == heapBase) {
+						rc = THROW_NPE;
+						goto done;
+					}
+					/* Extract the heap address from the base object of the heap segment which is
+					 * created by MemorySegment.ofArray(), MemorySegment.ofBuffer() or other methods
+					 * (which only returns part of the heap array) in OpenJDK.
+					 *
+					 * Note:
+					 * 1) Only MemorySegment objects representing the on-heap arrays reach here.
+					 * 2) Given the default offset of the original heap array is 8 bytes intended
+					 *    for the header, 8 bytes is subtracted from the offset of the passed-in
+					 *    heap segment so as to skip over the header (8 bytes) of the heap array
+					 *    to the content of the array against the offset.
+					 */
+					pointerValues[i] = (UDATA)(intptr_t)J9JAVAARRAY_EA(_currentThread, heapBase, heapOffset - 8, U_64);
```

Comment:
I don't think `8` should be hard-coded here, but perhaps more problematic is the construction of a pointer within an object on the heap. The GC may have difficulty with pointers into the heap, but not at the beginning of an object.

Reply:
The base object of the heap argument defined in the OpenJDK holds the address somewhere in the memory in which case the only way to access the array content is to determine how long the header bytes in the base object (which totally depends on the underlying implementation in OpenJDK as we have no idea of how the base object is defined to hold a array address based on the code at https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/src/java.base/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java#L62

That's why we have to skip the base bject header which I agree can't be hard-coded. So I introduced a constant `J9_FFI_DOWNCALL_HEAP_ARGUMENT_HEADER_BYTES` to accommodate any update for the base object in OpenJDK.

> The GC may have difficulty with pointers into the heap,

There is no GC related operations in the critical downcall with the heap argument as explained in OpenJDK.

Reply:
As discussed with @dmitripivkine as to how the Object model defined in GC, it seems fine to let `J9JAVAARRAY_EA` to locate the starting address of heap array by skipping the header assuming the array is contiguous in the case of `optthruput/optavgpause/gencon` but might be incorrect in other GC policies such case `Balanced/Metronome` when the array is discontiguous.   So, we need to figure out ways to cover both contiguous and discontiguous cases in our code to locate the heap array elements.

@TobiAjila, any input on this in terms of the discontiguous case with the heap array?

Reply:
I double-checked the implementation in the OpenJDK extension as how the heap array is created. e.g. an int array:
https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/src/java.base/share/classes/java/lang/foreign/MemorySegment.java#L1394
```
static MemorySegment ofArray(int[] intArray) {
        return SegmentFactories.fromArray(intArray); <----
    }
```
https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/src/java.base/share/classes/jdk/internal/foreign/SegmentFactories.java#L99
```
   public static MemorySegment fromArray(int[] arr) {
        ensureInitialized();
        Objects.requireNonNull(arr);
        long byteSize = (long)arr.length * Utils.BaseAndScale.INT.scale();
        return new OfInt(Utils.BaseAndScale.INT.base(),   <----- the default offset
                 arr, <---------------------------------------------- the heap array 
                 byteSize, false,
                MemorySessionImpl.createHeap(arr));
    }
```
https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/openj9/src/java.base/share/classes/jdk/internal/foreign/Utils.java#L306
```
 public record BaseAndScale(int base, long scale) {
...
        public static final BaseAndScale INT =
                new BaseAndScale(Unsafe.ARRAY_INT_BASE_OFFSET, Unsafe.ARRAY_INT_INDEX_SCALE);
```
https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/src/jdk.unsupported/share/classes/sun/misc/Unsafe.java#L758
```
    /** The value of {@code arrayBaseOffset(int[].class)} */
    public static final int ARRAY_INT_BASE_OFFSET = jdk.internal.misc.Unsafe.ARRAY_INT_BASE_OFFSET;
```
https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/src/java.base/share/classes/jdk/internal/misc/Unsafe.java#L1199
```
    /** The value of {@code arrayBaseOffset(int[].class)} */
    public static final int ARRAY_INT_BASE_OFFSET = theUnsafe.arrayBaseOffset(int[].class);
```
https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/src/java.base/share/classes/jdk/internal/misc/Unsafe.java#L1173
```
    public int arrayBaseOffset(Class<?> arrayClass) {
        if (arrayClass == null) {
            throw new NullPointerException();
        }

        return arrayBaseOffset0(arrayClass); <------
    }

private native int arrayBaseOffset0(Class<?> arrayClass);
```
https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/src/java.base/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java#L178C4-L183C1
```
 public static final class OfInt extends HeapMemorySegmentImpl {

        OfInt(long offset, Object base, long length, boolean readOnly, MemorySessionImpl session) {
            super(offset, base, length, readOnly, session);
        }

    HeapMemorySegmentImpl(long offset, Object base, long length, boolean readOnly, MemorySessionImpl session) {
        super(length, readOnly, session);
        this.offset = offset;
        this.base = base;
    }
```
So default offset of the heap array is `Utils.BaseAndScale.INT.base()` which is `ARRAY_INT_BASE_OFFSET` while the base object is the passed-in heap array `int[] arr` which follows the rules of the Object model defined by GC in OpenJ9.

According to the explanation on `Unsafe.ARRAY_INT_BASE_OFFSET` on Hotspot at https://stackoverflow.com/questions/12714064/hotspot-jvm-array-allocation:
```
Arrays are allocated as contiguous blocks. The size can be derived by using the class [sun.misc.Unsafe](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/7-b147/sun/misc/Unsafe.java#Unsafe) (Some great tutorial [here](http://robaustin.wikidot.com/how-to-write-to-direct-memory-locations-in-java)), which gives you native access to raw memory.
...
Due to implementation of the hotspot-jvm, all objects are aligned to 8 or 4 bytes (depending on your platform: AMD64 or x86_32), so the actual size of an array is increased to a multiple of 8 or 4 bytes.
```
So it looks like the heap array created in OpenJDK is assumed to be allocated on a contiguous memory with the default offset is 8 bytes (for the header + length depending on the implementation in OpenJ9 in terms of `Unsafe.ARRAY_INT_BASE_OFFSET`).

Reply:
If array must be contiguous you can use `J9JAVAARRAYCONTIGUOUS_EA` explicitly. Also it is reasonable to add assertion with `J9ISCONTIGUOUSARRAY_VM` check to be sure an array is in Contiguous form. I guess TBD how we are going to handle Discontiguous cases.

Reply:
>So it looks like the heap array created in OpenJDK is assumed to be allocated on a contiguous memory with the default offset is 8-bytes aligned depending on the implementation in JVM (which is 8 bytes for the header + length as defined in GC/OpenJ9 in terms of Unsafe.ARRAY_INT_BASE_OFFSET).

Object header is a GC property, you can't assume its size. Indexable object header size depends on Compressed or Full-64 _header_ format. Also it is selected dynamically at JVM startup based on GC policy. Actual size in bytes is stored in `javaVM->contiguousIndexableHeaderSize`

Reply:
Updated the code there against the suggested above.
```
					Assert_VM_true(J9ISCONTIGUOUSARRAY(_currentThread, heapBase));
					pointerValues[i] = (UDATA)(intptr_t)J9JAVAARRAYCONTIGUOUS_EA(
							_currentThread,
							heapBase,
							heapOffset - J9VMTHREAD_CONTIGUOUS_INDEXABLE_HEADER_SIZE(_currentThread),
							U_8);
```

Reply:
> If array must be contiguous you can use `J9JAVAARRAYCONTIGUOUS_EA` explicitly. Also it is reasonable to add assertion with `J9ISCONTIGUOUSARRAY_VM` check to be sure an array is in Contiguous form. I guess TBD how we are going to handle Discontiguous cases.

Yes, it must a contiguous array based on the Jtreg tests e.g. at https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/ff0372607f0412883d175e50595f854888300ef9/test/jdk/java/foreign/critical/libCritical.c#L56
```
// copy bytes into heap array
EXPORT void test_allow_heap_void(unsigned char* heapArr, unsigned char* nativeArr, int numBytes) {
    for (int i = 0; i < numBytes; i++) {
        heapArr[i] = nativeArr[i];   <--------
    }
}
```

Reply:
There is another implementation detail: there are two (equivalent for gencon/optthruput/optavgpause) ways to get address of the array data (first element) - get "object_address + header_size" or "get address of the first element" using macro. First way assumes object header is followed by array data. This assumption is not correct for coming "off-heap" implementation for Balanced, when Indexable object header and actual contiguous array data might be located separately. So, macro is going to return proper pointer to array data. It is irrelevant for gencon/optthruput/optavgpause now, but might help to avoid another code modification in order to support Balanced with "off-heap" on. It makes sense to have code ready.

Reply:
> So, macro is going to return proper pointer to array data. It is irrelevant for gencon/optthruput/optavgpause now,

 I am not sure whether there is any existing macro supporting this operation to ensure it works for both cases.

Reply:
`J9JAVAARRAYCONTIGUOUS_EA_VM` is going to return correct address for "off-heap" case as well, index 0 points to beginning of array data

Reply:
I didn't see any difference between `J9JAVAARRAYCONTIGUOUS_EA` and `J9JAVAARRAYCONTIGUOUS_EA_VM`
```
#define J9JAVAARRAYCONTIGUOUS_EA(vmThread, array, index, elemType) \
	(&((elemType*)((((UDATA)(array)) + (vmThread)->contiguousIndexableHeaderSize)))[(index)])

#define J9JAVAARRAYCONTIGUOUS_EA_VM(javaVM, array, index, elemType) \
	(&((elemType*)((((UDATA)(array)) + (javaVM)->contiguousIndexableHeaderSize)))[(index)])
```
If `J9JAVAARRAYCONTIGUOUS_EA_VM` works fine, then `J9JAVAARRAYCONTIGUOUS_EA` should be fine in both cases.

Reply:
right, both of these macro can be used.

Reply:
I'd like to use heapOffset = segment.unsafeGetOffset() - Unsafe.ARRAY_BYTE_BASE_OFFSET to compute the actual offset in the heap array in our java code. So there is no need to subtract the header from the heapOffset in the interpreter in native .e.g.
```
pointerValues[i] = (UDATA)(intptr_t)J9JAVAARRAYCONTIGUOUS_EA(_currentThread, heapBase, heapOffset, U_8);
```
But I have to compile a build to see whether it really works as expected in all primitive arrays.

Reply:
It works good with the proposed code above and has been verified in tests.

Reply:
WRT to on-heap arrays being passed to a native function. There is no way to support this with discontiguous arrays. Also, there is no graceful way to fail out in this case. So the only option AFAIK is to do something sub-optimal in these hopefully "rare" cases (this is something we will need to document). 

The basic idea is that the FLA code will detect an on-heap array arg being passed out, if it is discontiguous, we will allocate an off-heap array and copy the contents of the onheap array out. On the function return we will do the opposite.

Reply:
We already do something like this for JNI array critical, that can be a starting point.

Reply:
As discussed with @tajila offline, we'd like to let this PR handle the contiguous arrays while have the discontiguous case addressed in a separate PR due to the implementation specifics to be sorted out there.

Reply:
Probably it would be better to end up with an exception for the moment in the case of dis-contiguous arrays.

@dmitripivkine, any input about that?

Reply:
I think an exception would be much better than an assertion.

Reply:
Added an exception `ArrayStoreException` to capture any unexpected behavior when copying the discontiguous array in our case.